File: lib/fun_ci/ansi.rb
# frozen_string_literal: true

module FunCi
  module Ansi
    RESET = "\e[0m"

    def self.green(text)      = "\e[32m#{text}#{RESET}"
    def self.bold_green(text)  = "\e[1;32m#{text}#{RESET}"
    def self.bold_red(text)    = "\e[1;31m#{text}#{RESET}"
    def self.bold_yellow(text) = "\e[1;33m#{text}#{RESET}"
    def self.cyan(text)        = "\e[36m#{text}#{RESET}"
    def self.bold_cyan(text)   = "\e[1;36m#{text}#{RESET}"
    def self.dim(text)         = "\e[2m#{text}#{RESET}"
    def self.bold_white(text)  = "\e[1;37m#{text}#{RESET}"
    def self.white(text)       = "\e[37m#{text}#{RESET}"
    def self.yellow(text)      = "\e[33m#{text}#{RESET}"
    def self.dim_green(text)   = "\e[2;32m#{text}#{RESET}"
    def self.dim_red(text)     = "\e[2;31m#{text}#{RESET}"
    def self.orange(text)      = "\e[38;5;208m#{text}#{RESET}"
    def self.dark_red(text)    = "\e[38;5;52m#{text}#{RESET}"
    def self.bg_charcoal(text) = "\e[48;5;236m#{text}#{RESET}"
    def self.bg_dark_red(text)    = "\e[48;5;124m#{text}#{RESET}"
    def self.bg_orange(text)      = "\e[48;5;166m#{text}#{RESET}"
    def self.bg_very_dark_red(text) = "\e[48;5;52m#{text}#{RESET}"
    def self.bg_dark_green(text)  = "\e[48;5;22m#{text}#{RESET}"
    def self.bg_dark_yellow(text) = "\e[48;5;58m#{text}#{RESET}"

    def self.strip(text)
      text.gsub(/\e\[[0-9;]*[A-Za-z]/, "")
    end
  end
end

File: renderer/src/ansi.rs
//! SGR colouring, byte-compatible with the 1.x Ruby `Ansi` module.

pub const RESET: &str = "\u{1b}[0m";
pub const GREEN: &str = "32";
pub const BOLD_GREEN: &str = "1;32";
pub const BOLD_RED: &str = "1;31";
pub const BOLD_YELLOW: &str = "1;33";
pub const CYAN: &str = "36";
pub const BOLD_CYAN: &str = "1;36";
pub const DIM: &str = "2";

/// `text` in the SGR `code`, then a reset.
#[must_use]
pub fn paint(code: &str, text: &str) -> String {
    format!("\u{1b}[{code}m{text}{RESET}")
}

/// `text` with every CSI escape (`ESC [ digits/semicolons letter`) removed.
#[must_use]
pub fn strip(text: &str) -> String {
    let mut parts = text.split('\u{1b}');
    let head = parts.next().unwrap_or_default().to_string();
    head + &parts.map(after_escape).collect::<String>()
}

fn after_escape(part: &str) -> String {
    let params = part.strip_prefix('[').map(|rest| rest.trim_start_matches(|c: char| c.is_ascii_digit() || c == ';'));
    match params.filter(|rest| rest.starts_with(|c: char| c.is_ascii_alphabetic())) {
        Some(rest) => rest[1..].to_string(),
        None => format!("\u{1b}{part}"),
    }
}
